### PR TITLE
Fix k3d upgrade tests

### PR DESCRIPTION
**Pull request by jeremyharisch**
**Description**

Changes proposed in this pull request:
- Enhance Gardener test functions to include a custom directory path
- Re-Introduce `kyma-upgrade-gardener-kyma2-minor-versions` and fix reconiler test issues -> Pipeline still fails with `in-cluster event should be delivered`, but that will be fixed in a follow up issue for the evening team.
- Plus fix the following pipelines:
  - kyma-upgrade-gardener-kyma2-to-main
  - kyma-upgrade-k3d-kyma2-to-main
  - pre-main-reconciler-cli-kyma-prev-to-last-release-upgrade-k3d
  - pre-main-reconciler-cli-integration-kyma-latest-k3d

**Related issue(s)**

https://github.com/kyma-project/kyma/issues/16430

### CODE REVIEW

File: prow/scripts/lib/gardener/azure.sh [hunk 1]
```diff
@@ -283,19 +283,49 @@ gardener::test_fast_integration_kyma() {
...
+    log::info "Running pre-upgrade Kyma Fast Integration tests - Azure"
+
+    log::info "Checking Kyma fast-integration tests optional arguments"
+        while getopts ":d:" opt; do
+            case $opt in
+                d)
+                    kymaDirectory="$OPTARG"
+                    log::info "Kyma Source Directory: ${kymaDirectory}"
+                    ;;
+                \?)
```
Comment (tobiscr) [inline prow/scripts/lib/gardener/azure.sh:296]: Would recommend to use `*)` as fallback

File: prow/scripts/lib/utils.sh [hunk 2]
```diff
@@ -871,3 +871,25 @@ function utils::install_helm {
     popd || exit
     eval "${settings}"
 }
+
+function utils::get_kyma_fast_integration_dir {
+  log::info "### GET KYMA FAST INTEGRATION DIR"
```
Comment (tobiscr) [inline prow/scripts/lib/utils.sh:876]:
```suggestion
```

File: prow/scripts/lib/utils.sh [hunk 3]
```diff
@@ -871,3 +871,25 @@ function utils::install_helm {
...
+          fi
+          shift 2
+          ;;
+        *)
+          shift
+          ;;
+    esac
+  done
+
+  log::info "### Kyma Fast-integration-Test-Directory $kymaDirectory"
```
Comment (tobiscr) [inline prow/scripts/lib/utils.sh:893]:
```suggestion
```

File: prow/scripts/lib/utils.sh [hunk 4]
```diff
@@ -871,3 +871,25 @@ function utils::install_helm {
...
+  echo "$kymaDirectory"
```
Comment (tobiscr) [inline prow/scripts/lib/utils.sh:894]:
```suggestion
  echo -n "$kymaDirectory"
```

File: prow/scripts/cluster-integration/kyma-upgrade-gardener-kyma2-minor-versions.sh [hunk 5]
```diff
@@ -54,11 +54,11 @@ function deploy_base() {
   log::info "### Installing Kyma $KYMA_SOURCE"
 
   # uses previously set KYMA_SOURCE
-  kyma2_install_dir="$KYMA_SOURCES_DIR/kyma2.$((valid_minor_release_count))"
+  kyma2_install_dir="$KYMA_SOURCES_DIR/$KYMA_SOURCE"
   kyma::deploy_kyma -s "$KYMA_SOURCE" -d "$kyma2_install_dir" -u "true"
 
-  log::info "### Run pre-upgrade tests"
-  gardener::pre_upgrade_test_fast_integration_kyma
+  log::info "### test directory: $kyma2_install_dir/tests/fast-integration"
```
Comment (tobiscr) [inline prow/scripts/cluster-integration/kyma-upgrade-gardener-kyma2-minor-versions.sh:60]:
```suggestion
  log::info "### test directory: '$kyma2_install_dir/tests/fast-integration'"
```

File: prow/scripts/lib/gardener/aws.sh [hunk 6]
```diff
@@ -133,19 +133,23 @@ gardener::test_fast_integration_kyma() {
 }
 
 gardener::pre_upgrade_test_fast_integration_kyma() {
-    log::info "Running pre-upgrade Kyma Fast Integration tests"
+    log::info "Running pre-upgrade Kyma Fast Integration tests - AWS"
 
-    pushd /home/prow/go/src/github.com/kyma-project/kyma/tests/fast-integration
+    kymaDirectory="${utils::get_kyma_fast_integration_dir $@}"
+    log::info "Switching directory to $kymaDirectory"
```
Comment (tobiscr) [inline prow/scripts/lib/gardener/aws.sh:139]:
```suggestion
    log::info "Switching directory to '$kymaDirectory'"
```

File: prow/scripts/lib/gardener/aws.sh [hunk 7]
```diff
@@ -133,19 +133,23 @@ gardener::test_fast_integration_kyma() {
...
     log::success "Tests completed"
 }
 
 gardener::post_upgrade_test_fast_integration_kyma() {
-    log::info "Running post-upgrade Kyma Fast Integration tests"
+    log::info "Running post-upgrade Kyma Fast Integration tests - AWS"
 
-    pushd /home/prow/go/src/github.com/kyma-project/kyma/tests/fast-integration
+    kymaDirectory="${utils::get_kyma_fast_integration_dir $@}"
+    log::info "Switching directory to $kymaDirectory"
```
Comment (tobiscr) [inline prow/scripts/lib/gardener/aws.sh:151]:
```suggestion
    log::info "Switching directory to '$kymaDirectory'"
```

File: prow/scripts/lib/gardener/azure.sh [hunk 8]
```diff
@@ -283,19 +283,23 @@ gardener::test_fast_integration_kyma() {
 }
 
 gardener::pre_upgrade_test_fast_integration_kyma() {
-    log::info "Running pre-upgrade Kyma Fast Integration tests"
+    log::info "Running pre-upgrade Kyma Fast Integration tests - Azure"
 
-    pushd /home/prow/go/src/github.com/kyma-project/kyma/tests/fast-integration
+    kymaDirectory="${utils::get_kyma_fast_integration_dir $@}"
+    log::info "Switching directory to $kymaDirectory"
```
Comment (tobiscr) [inline prow/scripts/lib/gardener/azure.sh:289]:
```suggestion
    log::info "Switching directory to '$kymaDirectory'"
```

File: prow/scripts/lib/gardener/azure.sh [hunk 9]
```diff
@@ -283,19 +283,23 @@ gardener::test_fast_integration_kyma() {
...
     log::success "Tests completed"
 }
 
 gardener::post_upgrade_test_fast_integration_kyma() {
-    log::info "Running post-upgrade Kyma Fast Integration tests"
+    log::info "Running post-upgrade Kyma Fast Integration tests - Azure"
 
-    pushd /home/prow/go/src/github.com/kyma-project/kyma/tests/fast-integration
+    kymaDirectory="${utils::get_kyma_fast_integration_dir $@}"
+    log::info "Switching directory to $kymaDirectory"
```
Comment (tobiscr) [inline prow/scripts/lib/gardener/azure.sh:301]:
```suggestion
    log::info "Switching directory to '$kymaDirectory'"
```

File: prow/scripts/lib/gardener/gcp.sh [hunk 10]
```diff
@@ -135,19 +135,23 @@ gardener::test_fast_integration_kyma() {
 }
 
 gardener::pre_upgrade_test_fast_integration_kyma() {
-    log::info "Running pre-upgrade Kyma Fast Integration tests"
+    log::info "Running pre-upgrade Kyma Fast Integration tests - GCP"
 
-    pushd /home/prow/go/src/github.com/kyma-project/kyma/tests/fast-integration
+    kymaDirectory="${utils::get_kyma_fast_integration_dir $@}"
+    log::info "Switching directory to $kymaDirectory"
```
Comment (tobiscr) [inline prow/scripts/lib/gardener/gcp.sh:141]:
```suggestion
    log::info "Switching directory to '$kymaDirectory'"
```

File: prow/scripts/lib/gardener/gcp.sh [hunk 11]
```diff
@@ -135,19 +135,23 @@ gardener::test_fast_integration_kyma() {
...
     log::success "Tests completed"
 }
 
 gardener::post_upgrade_test_fast_integration_kyma() {
-    log::info "Running post-upgrade Kyma Fast Integration tests"
+    log::info "Running pre-upgrade Kyma Fast Integration tests - GCP"
 
-    pushd /home/prow/go/src/github.com/kyma-project/kyma/tests/fast-integration
+    kymaDirectory="${utils::get_kyma_fast_integration_dir $@}"
+    log::info "Switching directory to $kymaDirectory"
```
Comment (tobiscr) [inline prow/scripts/lib/gardener/gcp.sh:153]:
```suggestion
    log::info "Switching directory to '$kymaDirectory'"
```

**Review by tobiscr — CHANGES_REQUESTED** [review]

File: prow/scripts/lib/gardener/azure.sh [hunk 12]
```diff
@@ -283,19 +283,23 @@ gardener::test_fast_integration_kyma() {
 }
 
 gardener::pre_upgrade_test_fast_integration_kyma() {
-    log::info "Running pre-upgrade Kyma Fast Integration tests"
+    log::info "Running pre-upgrade Kyma Fast Integration tests - Azure"
 
-    pushd /home/prow/go/src/github.com/kyma-project/kyma/tests/fast-integration
+    kymaDirectory="${utils::get_kyma_fast_integration_dir $@}"
```
Comment (tobiscr) [inline prow/scripts/lib/gardener/azure.sh:288]:
```suggestion
    kymaDirectory="$(utils::get_kyma_fast_integration_dir $@)"
```

File: prow/scripts/lib/gardener/azure.sh [hunk 13]
```diff
@@ -283,19 +283,23 @@ gardener::test_fast_integration_kyma() {
...
 
     log::success "Tests completed"
 }
 
 gardener::post_upgrade_test_fast_integration_kyma() {
-    log::info "Running post-upgrade Kyma Fast Integration tests"
+    log::info "Running post-upgrade Kyma Fast Integration tests - Azure"
 
-    pushd /home/prow/go/src/github.com/kyma-project/kyma/tests/fast-integration
+    kymaDirectory="${utils::get_kyma_fast_integration_dir $@}"
```
Comment (tobiscr) [inline prow/scripts/lib/gardener/azure.sh:300]:
```suggestion
    kymaDirectory="$(utils::get_kyma_fast_integration_dir $@)"
```

File: prow/scripts/lib/gardener/gcp.sh [hunk 14]
```diff
@@ -135,19 +135,23 @@ gardener::test_fast_integration_kyma() {
 }
 
 gardener::pre_upgrade_test_fast_integration_kyma() {
-    log::info "Running pre-upgrade Kyma Fast Integration tests"
+    log::info "Running pre-upgrade Kyma Fast Integration tests - GCP"
 
-    pushd /home/prow/go/src/github.com/kyma-project/kyma/tests/fast-integration
+    kymaDirectory="${utils::get_kyma_fast_integration_dir $@}"
```
Comment (tobiscr) [inline prow/scripts/lib/gardener/gcp.sh:140]:
```suggestion
    kymaDirectory="$(utils::get_kyma_fast_integration_dir $@)"
```

File: prow/scripts/lib/gardener/gcp.sh [hunk 15]
```diff
@@ -135,19 +135,23 @@ gardener::test_fast_integration_kyma() {
...
 
     log::success "Tests completed"
 }
 
 gardener::post_upgrade_test_fast_integration_kyma() {
-    log::info "Running post-upgrade Kyma Fast Integration tests"
+    log::info "Running pre-upgrade Kyma Fast Integration tests - GCP"
 
-    pushd /home/prow/go/src/github.com/kyma-project/kyma/tests/fast-integration
+    kymaDirectory="${utils::get_kyma_fast_integration_dir $@}"
```
Comment (tobiscr) [inline prow/scripts/lib/gardener/gcp.sh:152]:
```suggestion
    kymaDirectory="$(utils::get_kyma_fast_integration_dir $@)"
```

File: prow/scripts/lib/gardener/aws.sh [hunk 16]
```diff
@@ -133,19 +133,23 @@ gardener::test_fast_integration_kyma() {
 }
 
 gardener::pre_upgrade_test_fast_integration_kyma() {
-    log::info "Running pre-upgrade Kyma Fast Integration tests"
+    log::info "Running pre-upgrade Kyma Fast Integration tests - AWS"
 
-    pushd /home/prow/go/src/github.com/kyma-project/kyma/tests/fast-integration
+    kymaDirectory="${utils::get_kyma_fast_integration_dir $@}"
```
Comment (tobiscr) [inline prow/scripts/lib/gardener/aws.sh:138]:
```suggestion
    kymaDirectory="$(utils::get_kyma_fast_integration_dir $@)"
```

File: prow/scripts/lib/gardener/aws.sh [hunk 17]
```diff
@@ -133,19 +133,23 @@ gardener::test_fast_integration_kyma() {
...
 
     log::success "Tests completed"
 }
 
 gardener::post_upgrade_test_fast_integration_kyma() {
-    log::info "Running post-upgrade Kyma Fast Integration tests"
+    log::info "Running post-upgrade Kyma Fast Integration tests - AWS"
 
-    pushd /home/prow/go/src/github.com/kyma-project/kyma/tests/fast-integration
+    kymaDirectory="${utils::get_kyma_fast_integration_dir $@}"
```
Comment (tobiscr) [inline prow/scripts/lib/gardener/aws.sh:150]:
```suggestion
    kymaDirectory="$(utils::get_kyma_fast_integration_dir $@)"
```

**Review by tobiscr — CHANGES_REQUESTED** [review]

File: prow/scripts/cluster-integration/kyma-upgrade-gardener-kyma2-minor-versions.sh [hunk 18]
```diff
@@ -83,13 +83,15 @@ function upgrade() {
...
+    gardener::post_upgrade_test_fast_integration_kyma -d "$kyma2_install_dir/tests/fast-integration"
 
     log::info "### waiting some time to finish cleanups"
     sleep 60
 
     log::info "### Run pre-upgrade tests again to validate component removal"
-    gardener::pre_upgrade_test_fast_integration_kyma
+    export KYMA_SOURCE="${minor_release_versions[$((i))]}"
+    kyma2_install_dir="$KYMA_SOURCES_DIR/kyma2.$i"
+    gardener::pre_upgrade_test_fast_integration_kyma -d "$kyma2_install_dir"/tests/fast-integration
```
Comment (tobiscr) [inline prow/scripts/cluster-integration/kyma-upgrade-gardener-kyma2-minor-versions.sh:94]:
```suggestion
    gardener::pre_upgrade_test_fast_integration_kyma -d "${kyma2_install_dir}/tests/fast-integration"
```

File: prow/scripts/cluster-integration/kyma-upgrade-gardener-kyma2-to-main.sh [hunk 19]
```diff
@@ -110,13 +110,13 @@ set +x
...
-kyma::deploy_kyma -s "$KYMA_SOURCE" -d "$kyma2_install_dir" -u "true"
+# kyma::deploy_kyma -s "$KYMA_SOURCE" -d "$kyma2_install_dir" -u "true"
 
 # generate pod-security-policy list in json
 utils::save_psp_list "${ARTIFACTS}/kyma-psp.json"
 
-log::info "### Run pre-upgrade tests"
-gardener::pre_upgrade_test_fast_integration_kyma
+# Pre-Upgrade Tests
+gardener::pre_upgrade_test_fast_integration_kyma -d kyma2/"$KYMA_SOURCE"/tests/fast-integration
```
Comment (tobiscr) [inline prow/scripts/cluster-integration/kyma-upgrade-gardener-kyma2-to-main.sh:119]:
```suggestion
gardener::pre_upgrade_test_fast_integration_kyma -d "kyma2/${KYMA_SOURCE}/tests/fast-integration"
```

**Review by tobiscr — DISMISSED** [review]

File: prow/scripts/lib/gardener/gcp.sh [hunk 20]
```diff
@@ -135,19 +135,23 @@ gardener::test_fast_integration_kyma() {
...
+    pushd "$kymaDirectory"
     make ci-pre-upgrade
     popd
 
     log::success "Tests completed"
 }
 
 gardener::post_upgrade_test_fast_integration_kyma() {
-    log::info "Running post-upgrade Kyma Fast Integration tests"
+    log::info "Running pre-upgrade Kyma Fast Integration tests - GCP"
```
Comment (ruanxin) [inline prow/scripts/lib/gardener/gcp.sh:150]:
```suggestion
    log::info "Running post-upgrade Kyma Fast Integration tests - GCP"
```

**Review by ruanxin — APPROVED** [review]

**Review by tobiscr — APPROVED** [review]

**Review by adityabhatia — APPROVED** [review]

**Review by Ressetkk — APPROVED** [review]

**Pull request merged**